### PR TITLE
many: add interface-specific constraints for home and camera interfaces

### PR DESCRIPTION
**Pull request by olivercalder**
Add support for fields in `Constraints`, `RuleConstraints`, and `RuleConstraintsPatch` to vary according to the interface associated with those constraints.

To facilitate this, create a new `InterfaceSpecificConstraints` interface which is embedded in all of these constraints types, since they all have the same interface-specific fields for a given interface. Also, add a new `ConstraintsJSON` type, which is a `map[string]json.RawMessage`, to be used as an intermediate form through which to marshal/unmarshal these constraints.

To demonstrate these constraints working properly, implement both `InterfaceSpecificConstraintsHome` and `InterfaceSpecificConstraintsCamera`.

Since lots of the the internal structures still assume that `PathPattern`s are used, `InterfaceSpecificConstraints` has a `PathPattern` method. For `InterfaceSpecificConstraintsHome`, this is the real path pattern from those constraints. For `InterfaceSpecificConstraintsCamera`, this returns a placeholder `PathPattern` representing `"/**"`, since camera replies/rules apply to any request for camera access, regardless of path.

`promptConstraints` also marshal to different fields depending on the interface (e.g. camera does not include the `"path"` field), but since requests all still contain paths, the structure of `promptConstraints` is not changed; instead, marshalling must be done according to a given interface.

This is a simplified approach with fewer changes than #16051, which this replaces. Since camera constraints are also implemented as part of this PR, this also supersedes #16052. When ready, #16053 should be rebased on this PR. This work is tracked internally by https://warthogs.atlassian.net/browse/SNAPDENG-35661.

### CODE REVIEW

File: interfaces/prompting/requestrules/requestrules.go [hunk 1]
```diff
@@ -233,12 +259,11 @@ func (rdb *RuleDB) load() (retErr error) {
...
 	if err != nil {
-		// TODO: store rules separately per-user, so a corrupted rule for one
-		// user can't impact rules for another user.
-		loadErr := fmt.Errorf("cannot read stored request rules: %w", err)
 		// Save the empty rule DB to disk to overwrite the previous one which
 		// could not be decoded.
-		return strutil.JoinErrors(loadErr, rdb.save())
+		// TODO: store rules separately per-user, so a corrupted rule for one
+		// user can't impact rules for another user.
+		return strutil.JoinErrors(err, rdb.save())
```
Comment (pedronis) [inline interfaces/prompting/requestrules/requestrules.go:266]: why the error changes

File: interfaces/prompting/requestrules/requestrules.go [hunk 2]
```diff
@@ -257,7 +282,7 @@ func (rdb *RuleDB) load() (retErr error) {
 		expired, err := rule.validate(at)
 		if err != nil {
 			// we're loading previously saved rules, so this should not happen
-			errInvalid = fmt.Errorf("internal error: %w", err)
+			errInvalid = err
```
Comment (pedronis) [inline interfaces/prompting/requestrules/requestrules.go:285]: same question

File: interfaces/prompting/constraints.go [hunk 3]
```diff
@@ -31,12 +32,206 @@ import (
...
+
+func (constraints *InterfaceSpecificConstraintsHome) parseJSON(constraintsJSON ConstraintsJSON) error {
+	// Expect fields: "path-pattern"
+	pathPatternJSON, ok := constraintsJSON["path-pattern"]
+	if !ok {
+		return prompting_errors.NewInvalidPathPatternError("", "no path pattern")
+	}
+	var pathPattern patterns.PathPattern
+	if err := pathPattern.UnmarshalJSON(pathPatternJSON); err != nil {
+		// XXX: should this be wrapped in a "cannot decode ..." message like in daemon?
```
Comment (pedronis) [inline interfaces/prompting/constraints.go]: possibly yes

File: interfaces/prompting/constraints.go [hunk 4]
```diff
@@ -289,6 +595,10 @@ func (pm PermissionMap) toRulePermissionMap(iface string, at At) (RulePermission
 			invalidPerms = append(invalidPerms, perm)
 			continue
 		}
+		if entry == nil {
+			// treat permissions with nil entries as if they were not present
+			continue
```
Comment (pedronis) [inline interfaces/prompting/constraints.go:597]: this code is new but the tests don't seem to hit it

File: interfaces/prompting/constraints.go [hunk 5]
```diff
@@ -46,17 +241,26 @@ type Constraints struct {
...
 		// Error should not occur, since it was parsed internally
-		return false, prompting_errors.NewInvalidPathPatternError(c.PathPattern.String(), err.Error())
+		return false, prompting_errors.NewInvalidPathPatternError(c.InterfaceSpecific.PathPattern().String(), err.Error())
 	}
 	return match, nil
 }
 
+// PathPattern returns the PathPattern provided by the interface-specific
+// constraints.
+func (c *Constraints) PathPattern() *patterns.PathPattern {
```
Comment (pedronis) [inline interfaces/prompting/constraints.go:259]: this is not tested

File: interfaces/prompting/constraints.go [hunk 6]
```diff
@@ -75,89 +279,154 @@ func (c *Constraints) ContainPermissions(permissions []string) bool {
...
-// ReplyConstraints hold information about the applicability of a reply to
-// particular paths and permissions. Upon receiving the reply, snapd converts
-// ReplyConstraints to Constraints.
-type ReplyConstraints struct {
-	PathPattern *patterns.PathPattern `json:"path-pattern"`
-	Permissions []string              `json:"permissions"`
+// PathPattern returns the PathPattern provided by the interface-specific
+// constraints.
+func (c *RuleConstraints) PathPattern() *patterns.PathPattern {
+	return c.InterfaceSpecific.PathPattern()
```
Comment (pedronis) [inline interfaces/prompting/constraints.go]: same

File: interfaces/prompting/constraints.go [hunk 7]
```diff
@@ -31,12 +32,206 @@ import (
...
+	// patch returns a new InterfaceSpecificConstraints with the receiver used
+	// to patch the given existing constraints.
+	patch(existing InterfaceSpecificConstraints) InterfaceSpecificConstraints
+}
+
+// ParseInterfaceSpecificConstraints parses the given constraints from json
+// according to the given interface. If isPatch is true, then the constraints
+// should be treated as a patch, where fields may be omitted to indicate that
+// they should be left unchanged from the rule which will be patched.
+func ParseInterfaceSpecificConstraints(iface string, constraintsJSON ConstraintsJSON, isPatch bool) (InterfaceSpecificConstraints, error) {
```
Comment (pedronis) [inline interfaces/prompting/constraints.go]: is this used in other packages? can we keep unexported and test it via export_test.go?

File: interfaces/prompting/requestprompts/requestprompts.go [hunk 8]
```diff
@@ -215,6 +207,47 @@ type promptConstraints struct {
...
+type promptConstraintsJSONCamera struct {
+	RequestedPermissions []string `json:"requested-permissions"`
+	AvailablePermissions []string `json:"available-permissions"`
+}
+
+func (pc *promptConstraints) MarshalJSON() ([]byte, error) {
+	panic("programmer error: cannot marshal promptConstraints directly; must use MarshalForInterface with a given interface")
+}
+
+func (pc *promptConstraints) MarshalForInterface(iface string) ([]byte, error) {
```
Comment (pedronis) [inline interfaces/prompting/requestprompts/requestprompts.go]: does this need to be exported?

File: interfaces/prompting/constraints.go (hunk 7 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go]: It can be unexported

File: interfaces/prompting/constraints.go (hunk 4 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go:597]: Yes, this is an edge case I spotted which I believe has been a bug. You can add a null permission entry when creating rule directly by POSTing to `/v2/interfaces/requests/rules`, which could then cause a null pointer exception, so this fixes that. And yes I need to add a dedicated test for this.

File: interfaces/prompting/requestprompts/requestprompts.go (hunk 8 above)
Comment (olivercalder) [reply interfaces/prompting/requestprompts/requestprompts.go]: No, good catch, thanks. Force of habit from `MarshalJSON` methods

File: interfaces/prompting/requestrules/requestrules.go (hunk 1 above)
Comment (olivercalder) [reply interfaces/prompting/requestrules/requestrules.go:266]: Rather than first unmarshalling and then validating the result, unmarshalling and validation now happens in stages. So it may be the case that we can in fact read stored request rules, but those rules are invalid. And since there's not a "cannot read stored request rules" prefix wrapping the error below from `validate()` (and one wouldn't make sense there IMO), I think for consistency there shouldn't be one here either.

Basically, I want errors from these two potential error locations to look the same, since it's a bit arbitrary whether some particular validation happens to occur during unmarshalling or during `validate()`.

File: interfaces/prompting/requestrules/requestrules.go (hunk 2 above)
Comment (olivercalder) [reply interfaces/prompting/requestrules/requestrules.go:285]: If the rules on disk are corrupted in some way, then this isn't really an "internal error", it's a validation error just like those which could occur during unmarshalling. So the intention is to have errors from `validate()` look just like errors from unmarshalling, without different prefixes.

File: interfaces/prompting/constraints.go (hunk 4 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go:597]: I added more tests for these scenarios, either `null` permission entry from the API or from disk.

**Review by olivercalder — COMMENTED** [review]
Addressed some feedback. Thanks!

File: interfaces/prompting/constraints.go [hunk 9]
```diff
@@ -75,89 +279,154 @@ func (c *Constraints) ContainPermissions(permissions []string) bool {
...
+	var permissionMap RulePermissionMap
+	if err := json.Unmarshal(permissionsJSON, &permissionMap); err != nil {
+		// XXX: should this be wrapped in a "cannot decode ..." message like in daemon?
+		return nil, err
+	}
+	// Permissions must be validated later via ValidateForInterface to check if
+	// any/all permission entries are invalid or expired.
+	// XXX: should permissions be validated here while unmarshalling, rather
+	// than later, and leave the later checks to just be for expiration? That
+	// would better match the behavior of UnmarshalReplyConstraints.
```
Comment (pedronis) [inline interfaces/prompting/constraints.go]: are these changes of validation easy to do? same about the other similar XXXs?

should they be done in a follow up?

Comment (olivercalder) [reply interfaces/prompting/constraints.go]: Yeah the changes are not hard to do, they are just more overall changes. Could be in a follow-up. This might affect how we test error responses coming out of the daemon, so worth deciding on this before writing those tests, I think. So both this and the error testing could be follow-up PRs. Or both in this PR.

File: interfaces/prompting/requestprompts/requestprompts.go [hunk 10]
```diff
@@ -63,41 +63,33 @@ const (
...
-	Constraints  *promptConstraints
-	listenerReqs []*listener.Request
+	ID           prompting.IDType    `json:"id"`
+	Timestamp    time.Time           `json:"timestamp"`
+	Snap         string              `json:"snap"`
+	PID          int32               `json:"pid"`
+	Cgroup       string              `json:"cgroup"`
+	Interface    string              `json:"interface"`
+	Constraints  *promptConstraints  `json:"constraints"`
+	listenerReqs []*listener.Request `json:"-"`
```
Comment (andrewphelpsj) [inline interfaces/prompting/requestprompts/requestprompts.go]:
```suggestion
	listenerReqs []*listener.Request
```
struct tag isn't needed, since the field isn't exported

Comment (andrewphelpsj) [reply interfaces/prompting/requestprompts/requestprompts.go]: Actually, this struct doesn't need tags at all, I don't think? Since we define how it gets encoded, and we convert it to a `jsonPrompt`.

File: interfaces/prompting/constraints.go [hunk 11]
```diff
@@ -75,89 +279,154 @@ func (c *Constraints) ContainPermissions(permissions []string) bool {
...
 		return nil, prompting_errors.NewInvalidInterfaceError(iface, availableInterfaces())
 	}
-	if len(c.Permissions) == 0 {
+	permissionsJSON, ok := constraintsJSON["permissions"]
+	if !ok {
+		return nil, prompting_errors.NewPermissionsEmptyError(iface, availablePerms)
+	}
+	var permissionsList []string
+	if err := json.Unmarshal(permissionsJSON, &permissionsList); err != nil {
+		// XXX: should this be wrapped in a "cannot decode ..." message like in daemon?
```
Comment (bboozzoo) [inline interfaces/prompting/constraints.go]: that'd be useful I think

File: interfaces/prompting/constraints.go [hunk 12]
```diff
@@ -188,8 +457,45 @@ func (c *ReplyConstraints) ToConstraints(iface string, outcome OutcomeType, life
...
+	if err != nil {
+		return nil, err
+	}
+	constraints.InterfaceSpecific = interfaceSpecific
+
+	permissionsJSON, ok := constraintsJSON["permissions"]
+	if ok {
+		var permissionMap PermissionMap
+		if err := json.Unmarshal(permissionsJSON, &permissionMap); err != nil {
+			// XXX: should this be wrapped in a "cannot decode ..." message like in daemon?
```
Comment (bboozzoo) [inline interfaces/prompting/constraints.go]: same

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 13]
```diff
@@ -439,24 +439,32 @@ func (m *InterfacesRequestsManager) HandleReply(userID uint32, promptID promptin
...
 	}
 
 	// XXX: do we want to allow only replying to a select subset of permissions, and
 	// auto-deny the rest?
 	contained := constraints.ContainPermissions(prompt.Constraints.OutstandingPermissions())
 	if !contained {
+		// XXX: since we don't have a ReplyConstraints type anymore, we never
+		// expose the original list of permissions in the reply. Instead, we
+		// need to reconstruct it from the keys in the permission map. Thus,
+		// the permissions will no longer be in their original order.
```
Comment (bboozzoo) [inline overlord/ifacestate/apparmorprompting/prompting.go]: Referencing a type that's being removed only makes sense in the context of a patch or a PR. Once the new code ands and someone looks at it again they will have no idea how the long gone `ReplyConstraints` is relevant here.

```suggestion
		// We never expose the original list of permissions in the reply so, we
		// need to reconstruct it from the keys in the permission map. Thus,
		// the permissions will no longer be in their original order.
```

File: interfaces/prompting/constraints.go [hunk 14]
```diff
@@ -31,12 +32,206 @@ import (
...
+		return nil, prompting_errors.NewPermissionsEmptyError(iface, availablePerms)
+	}
+	var permissionMap PermissionMap
+	if err := json.Unmarshal(permissionsJSON, &permissionMap); err != nil {
+		// XXX: should this be wrapped in a "cannot decode ..." message like in daemon?
+		return nil, err
+	}
+	// Permissions must be validated later via ToRuleConstraints to check if
+	// any permission entries are invalid.
+	// XXX: should permissions be validated here while unmarshalling, rather
```
Comment (bboozzoo) [inline interfaces/prompting/constraints.go]: same question as @pedronis for UnmarshalRuleConstraints, perhaps convert to a `TODO` and open a followup PR?

File: interfaces/prompting/constraints.go [hunk 15]
```diff
@@ -31,12 +32,206 @@ import (
...
+func (constraints *InterfaceSpecificConstraintsHome) PathPattern() *patterns.PathPattern {
+	return constraints.Pattern
+}
+
+func (constraints *InterfaceSpecificConstraintsHome) patch(existing InterfaceSpecificConstraints) InterfaceSpecificConstraints {
+	newConstraints := &InterfaceSpecificConstraintsHome{}
+	if existing != nil {
+		// Should never attempt to patch nil existing constraints
+		existingHome, ok := existing.(*InterfaceSpecificConstraintsHome)
+		if ok {
```
Comment (andrewphelpsj) [inline interfaces/prompting/constraints.go:147]: should we do something if `ok` is false here?

File: interfaces/prompting/constraints.go [hunk 16]
```diff
@@ -31,12 +32,206 @@ import (
...
+	// fields are not treated as an error, but instead left empty.
+	parsePatchJSON(constraintsJSON ConstraintsJSON) error
+	// toJSON returns a ConstraintsJSON so the caller can add the permissions
+	// to it and then marshal the result into json.
+	toJSON() (ConstraintsJSON, error)
+	// PathPattern returns a path pattern which can be used to match incoming
+	// requests against these constraints.
+	// XXX: Not all interfaces care about path patterns. For those that don't,
+	// this should return a placeholder designed to match any path, such as /**
+	PathPattern() *patterns.PathPattern
```
Comment (andrewphelpsj) [inline interfaces/prompting/constraints.go]: Why exported when the others are not? Just for testing?

File: interfaces/prompting/constraints.go (hunk 14 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go]: Sounds good, thanks!

File: interfaces/prompting/constraints.go (hunk 9 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go]: I would potentially like to move the validation into the unmarshalling of the permission map itself, so any permission map we unmarshal or build should be valid. But yes, marking this as a TODO and I'll address it in a dedicated PR.

File: interfaces/prompting/constraints.go (hunk 3 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go]: I actually think this should be the responsibility of the caller, since we'd also want to wrap the above "no path pattern" error in the same message. But really this applies for all unmarshalling of all constraints types. I think the caller of `UnmarshalConstraints` should be responsible for wrapping any resulting error in a `cannot decode` message, since any unmarshalling errors relate to decoding the message.

File: interfaces/prompting/constraints.go (hunk 11 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go]: I think not actually, I think this should be the responsibility of the caller. I've added a commit to change this.

File: interfaces/prompting/constraints.go (hunk 12 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go]: same, moved this error message wrapping up to the caller of `Unmarshal...`, since that's what knows exactly where the json is coming from.

File: interfaces/prompting/requestprompts/requestprompts.go (hunk 10 above)
Comment (olivercalder) [reply interfaces/prompting/requestprompts/requestprompts.go]: Yeah, you're totally right. I think I wanted to leave them there to make it clear which fields mapped to which json tags, but that's really not necessary.

File: interfaces/prompting/constraints.go (hunk 15 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go:147]: It should never occur, so extra error checking shouldn't really be necessary in practice, and not returning an `error` makes usage simpler. But that's not necessarily a good enough reason... if you think it would be definitely safer to handle the `!ok` case, I can probably do that, as we don't want to have an `InterfaceSpecificConstraintsHome` with nil `Pattern`.

File: interfaces/prompting/constraints.go (hunk 16 above)
Comment (olivercalder) [reply interfaces/prompting/constraints.go]: Yeah good catch, this can be unexported. I was waffling about whether users of constraints should be expected to manually extract the `PathPattern` to use it if needed, vs having a method on the constraints types themselves, but settled on the latter.

File: overlord/ifacestate/apparmorprompting/prompting.go [hunk 17]
```diff
@@ -592,14 +603,26 @@ func (m *InterfacesRequestsManager) RuleWithID(userID uint32, ruleID prompting.I
...
+	// Lookup existing rule only so we can get its interface, so we can
+	// unmarshal the constraints patch from json.
+	origRule, err := m.rules.RuleWithID(userID, ruleID)
+	if err != nil {
+		return nil, err
+	}
+	constraintsPatch, err := prompting.UnmarshalRuleConstraintsPatch(origRule.Interface, constraintsPatchJSON)
+	if err != nil {
+		// XXX: should this say "... or deletion" like daemon does?
+		return nil, fmt.Errorf("cannot decode request body into request rule modification: %w", err)
```
Comment (pedronis) [inline overlord/ifacestate/apparmorprompting/prompting.go:623]: this seems fine, drop XXX?
